src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/BulkConnectionRetryWrapper.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright © 2024 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.salesforce.plugin.source.batch.util;
+
+import com.sforce.async.AsyncApiException;
+import com.sforce.async.BatchInfo;
+import com.sforce.async.BatchInfoList;
+import com.sforce.async.BulkConnection;
+import com.sforce.async.JobInfo;
+import com.sforce.ws.ConnectorConfig;
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import io.cdap.plugin.salesforce.plugin.source.batch.SalesforceBulkRecordReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+
+/**
+ * BulkConnectionRetryWrapper class to retry all the salesforce api calls in case of failure.
+ */
+public class BulkConnectionRetryWrapper {
+
+  private final BulkConnection bulkConnection;
+  private final RetryPolicy retryPolicy;
+  private static final Logger LOG = LoggerFactory.getLogger(BulkConnectionRetryWrapper.class);
+  private final boolean retryOnBackendError;
+  private final long maxRetryDuration;
+  private final int maxRetryCount;
+  private final long initialRetryDuration;
+
+  public BulkConnectionRetryWrapper(BulkConnection bulkConnection, boolean isRetryRequired,
+                                    long initialRetryDuration, long maxRetryDuration, int maxRetryCount) {
+    this.bulkConnection = bulkConnection;
+    this.retryOnBackendError = isRetryRequired;
+    this.initialRetryDuration = initialRetryDuration;
+    this.maxRetryDuration = maxRetryDuration;
+    this.maxRetryCount = maxRetryCount;
+    this.retryPolicy = SalesforceSplitUtil.getRetryPolicy(initialRetryDuration, maxRetryDuration, maxRetryCount);
+  }
+
+  public JobInfo createJob(JobInfo jobInfo) throws AsyncApiException {
+    if (!retryOnBackendError) {
```

Comment:
rather than adding if condition for each method, can we have a retry policy that dont try after failure. E.g maxAttempts = 1